Apps/TimelineApp/feature/user/src/main/java/uk/ac/cam/cares/jps/user/SensorSettingFragment.java

```diff
@@ -90,196 +91,236 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
     public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
 
+        sensorViewModel.checkRecordingStatusAndUpdateUI(requireContext());
+
+        setupRecyclerView(view);
+
+        binding.startRecordTv.setEnabled(true);
+        binding.startRecordTv.setOnClickListener(this::onRecordButtonClicked);
+
+        setupObservers();
+
+        setupUIInteractions();
+
+    }
+
+    /**
+     * Sets up the RecyclerView for displaying sensor items, initializes the adapter,
+     * and configures the layout and item decorations for the RecyclerView.
+     *
+     * @param view The root view of the fragment where the RecyclerView is located.
+     */
+    private void setupRecyclerView(View view) {
         RecyclerView recyclerView = view.findViewById(R.id.sensors_recycler_view);
         recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
-
         DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
                 LinearLayoutManager.VERTICAL);
         recyclerView.addItemDecoration(dividerItemDecoration);
 
-
-        // Initialize the adapter with an empty list first
-        adapter = new SensorAdapter(new ArrayList<>(), this, sensorViewModel);
+        adapter = new SensorAdapter(new ArrayList<>(), sensorViewModel);
         recyclerView.setAdapter(adapter);
+    }
 
-
+    /**
+     * Sets up observers to monitor changes in the sensor items, the toggle-all state, and
+     * the recording state. Updates the UI and RecyclerView based on the observed data.
+     */
+    @SuppressLint("NotifyDataSetChanged")
+    private void setupObservers() {
         sensorViewModel.getSensorItems().observe(getViewLifecycleOwner(), sensorItems -> {
             adapter.updateSensorItems(sensorItems);
             adapter.notifyDataSetChanged();
         });
+        sensorViewModel.getAllToggledOn().observe(getViewLifecycleOwner(), this::updateToggleAllState);
+        sensorViewModel.getIsRecording().observe(getViewLifecycleOwner(), this::updateRecordingUI);
```

Comment:
the control of start/stop recording is still weird. Please recheck logics in `updateRecordingUI()`, `startRecording()`, `stopRecording()` and `SensorViewModel.stopRecording()` and  `SensorViewModel.startRecording()`.

- It seems that `startRecording()` in fragment manages both start and stop recording logic and also some of the UI component control. It overlap with `stopRecording()` and `updateRecordingUI()`
- A more ideal way will be when click start/stop button, the fragment only update the `isRecording` state in ViewModel and the rest is done with this observer, which require modification in SensorViewModel as well